tests/lib/openstack_query/query_blocks/test_query_builder.py

```diff
@@ -42,61 +41,95 @@ def instance_fixture(mock_client_side_handlers, mock_server_side_handler):
     """
-    Tests that parse_where functions expectedly - where inputs valid
-    method finds and sets client_side_filter and server_side_filters internal attributes
+    Fixture to run parse where test cases
     """
 
-    # setup mocks
-    mock_kwargs = {"arg1": "val1", "arg2": "val2"}
-
     mock_client_side_handler = MagicMock()
     mock_client_filter_func = MagicMock()
     mock_client_side_handler.get_filter_func.return_value = mock_client_filter_func
-    mock_server_side_handler.get_filters.return_value = mock_server_side_filter
-
-    with patch(
-        "openstack_query.query_blocks.query_builder.QueryBuilder._get_preset_handler"
-    ) as mock_get_preset_handler:
-        mock_get_preset_handler.return_value = mock_client_side_handler
-        with patch.object(MockProperties, "get_prop_mapping") as mock_prop_func:
-            instance.parse_where(
-                MockQueryPresets.ITEM_1, MockProperties.PROP_1, mock_kwargs
-            )
-
-    mock_get_preset_handler.assert_called_once_with(
-        MockQueryPresets.ITEM_1, MockProperties.PROP_1
-    )
-    mock_client_side_handler.get_filter_func.assert_called_once_with(
-        preset=MockQueryPresets.ITEM_1,
-        prop=MockProperties.PROP_1,
-        prop_func=mock_prop_func.return_value,
-        filter_func_kwargs=mock_kwargs,
-    )
-    mock_server_side_handler.get_filters.assert_called_once_with(
-        preset=MockQueryPresets.ITEM_1,
-        prop=MockProperties.PROP_1,
-        params=mock_kwargs,
-    )
 
-    assert instance.client_side_filter == mock_client_filter_func
-    assert instance.server_side_filters == mock_server_side_filter
+    def _run_parse_where_test_case(mock_server_side_filters, mock_kwargs):
+        """
+        run parse where test case where get_filter_func returns different server-side-filters
+        """
+
+        expected_server_side_filter = mock_server_side_filters
+
+        # add_filters expects a list of server-side filters
+        # - convert to singleton list if a single filter is returned by get_filter_func
+        if mock_server_side_filters and not isinstance(mock_server_side_filters, list):
+            expected_server_side_filter = [mock_server_side_filters]
+
+        mock_server_side_handler.get_filters.return_value = mock_server_side_filters
+        with patch(
+            "openstack_query.query_blocks.query_builder.QueryBuilder._get_preset_handler"
+        ) as mock_get_preset_handler:
+            with patch(
+                "openstack_query.query_blocks.query_builder.QueryBuilder._add_filter"
+            ) as mock_add_filter:
+                mock_get_preset_handler.return_value = mock_client_side_handler
+                with patch.object(MockProperties, "get_prop_mapping") as mock_prop_func:
+                    instance.parse_where(
+                        MockQueryPresets.ITEM_1, MockProperties.PROP_1, mock_kwargs
+                    )
```

Comment:
Why do we need to patch so much stuff for `get_filter_func`, this shouts to me the method is doing way too much

It should be fairly pure, e.g. `get_filter_func` gets pre-set filter functions, rather than generating them (which can be a separate step)

Reply:
I'll fix this later